### PR TITLE
generate netCDF4 files from cdl by default

### LINKED ISSUE #725: Generate netCDF-4 format file from cdl files instead of netCDF-classic format files
# Current situation:

When a cdl file is provided to the compliance checker, a netCDF-classic file is created via `ncgen`. 

Code (`suite.py`, l.669ff):

``` python
        if '.cdl' in cdl_path:  # it's possible the filename doesn't have the .cdl extension
            ds_str = cdl_path.replace('.cdl', '.nc')
        else:
            ds_str = cdl_path + '.nc'
        subprocess.call(['ncgen', '-o', ds_str, cdl_path])
        return ds_str
```

# Problem

The cdl file might contain netCDF-4 features that do not exist in the netCDF-classic format -- e.g. additional [primitive data types](https://www.unidata.ucar.edu/software/netcdf/docs/netcdf_utilities_guide.html#cdl_data_types). The usage of one/some of the additional/new data types causes `ncgen` to fail when a netCDF-classic file is create from the cdl file.

# Reproduction of the issue

Please find attached to cdl files:

* [example_with_netcdf4_feature.txt](https://github.com/ioos/compliance-checker/files/4155871/example_with_netcdf4_feature.txt)
* [example_without_netcdf4_feature.txt](https://github.com/ioos/compliance-checker/files/4155872/example_without_netcdf4_feature.txt)

``` bash
# succeeds (create netCDF classic file)
ncgen example_without_netcdf4_feature.txt -o example_without_netcdf4_feature.nc
ncgen -k nc3 example_without_netcdf4_feature.txt -o example_without_netcdf4_feature.nc

# fails (create netCDF classic file from cdl file with netCDF4 features)
ncgen example_with_netcdf4_feature.txt -o example_with_netcdf4_feature.nc
ncgen -k nc3 example_with_netcdf4_feature.txt -o example_with_netcdf4_feature.nc

# solution that succeeds for file `example_with_netcdf4_feature.txt`
ncgen -k nc4 example_with_netcdf4_feature.txt -o example_with_netcdf4_feature.nc
```

# Solution

We generate a netCDF4 file instead of a netCDF-classic file by modifying the code from

``` python
subprocess.call(['ncgen', '-o', ds_str, cdl_path])
```

to

``` python
subprocess.call(['ncgen', '-k', 'nc4', '-o', ds_str, cdl_path])
```

If this modification is of interest, I can create a pull request for this.

### PR DESCRIPTION
**Pull request by neumannd**
fixes #725:

* generate netCDF4 file from cdl with ncgen
* first try gen netCDF4; fallback Classic
* new cdl test file with netCDF4 features
* two new test for netCDF4 file generation

Background: When a CDL file is provided, currently, a netCDF classic file is created by ncgen. However, a CDL file might contain netCDF4 features such as attributes/variables of type int64 (indicated by `LL` after numbers). Therefore, the `ncgen`-call that creates a netCDF file was modified to create a netCDF4 instead of a netCDF-classic file.

### CODE REVIEW

File: compliance_checker/tests/data/test_cdl_nc4_file.cdl [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+		lat:standard_name = "latitude" ;
+		lat:units = "degrees_north" ;
+		lat:axis = "Y" ;
+
+// global attributes:
+		:title = "A test file" ;
+		:history = "2020-02-04T18:00 manually create by Daniel Neumann, DKRZ" ;
+		:global_att_of_type_int = 2 ;
+		:global_att_of_type_int64 = 2LL ;
+		:Conventions = "CF-1.6" ;
```
Comment (daltonkell) [inline compliance_checker/tests/data/test_cdl_nc4_file.cdl]: We can now use CF-1.7, so let's put it to good use!

File: compliance_checker/tests/test_suite.py [hunk 2]
```diff
@@ -184,3 +197,12 @@ def test_standard_output_score_header(self):
...
 
+    def test_netCDF4_features(self):
+        """
+        Check if a proper netCDF4 file with netCDF4-datatypes is created.
+        """
+        # create and open dataset
+        ds = self.cs.load_dataset(static_files['netCDF4'])
+        # check if types are correct
+        assert isinstance(ds.global_att_of_type_int, np.int32)
+        assert isinstance(ds.global_att_of_type_int64, np.int64)
```
Comment (daltonkell) [inline compliance_checker/tests/test_suite.py]: While you're already doing type-checking for global attributes, it probably wouldn't hurt to type-check the actual variables as well.

File: compliance_checker/tests/data/test_cdl_nc4_file.cdl (hunk 1 above)
Comment (neumannd) [reply compliance_checker/tests/data/test_cdl_nc4_file.cdl]: I will update it accordingly.

File: compliance_checker/suite.py [hunk 3]
```diff
@@ -670,7 +670,21 @@ def generate_dataset(self, cdl_path):
...
+          # if not successfull, create netCDF classic file
+          print('netCDF-4 file could not be generated from cdl file ' +
+                cdl_path + '. Trying to create netCDF Classic file.')
+          iostat = subprocess.call(['ncgen', '-k', 'nc3', '-o', ds_str, cdl_path])
+          if iostat != 0:
+            # If neither a netCDF Classic nor a netCDF-4 file could be
+            # created, throw an error.
+            raise RuntimeError('`ncgen` could not generate netCDF file ' +
+                               '(classic and netCDF4) from provided cdl' +
+                               ' file: ' + cdl_path + '. Generation of ' +
```
[the remark this answers is not in the document]
Comment (neumannd) [reply compliance_checker/suite.py]: I will have a look into the error handling.

File: compliance_checker/tests/test_suite.py (hunk 2 above)
Comment (neumannd) [reply compliance_checker/tests/test_suite.py]: This was mainly to check whether the netCDF4 type `int64` is properly used. Should I do the same for a variable?

**Comment by neumannd** [comment]
I will re-create this pull request from the most recent code base next week.

**Comment by neumannd** [comment]
This pull request was superseded by #731 . This was the easiest solution to solve the conflicts.

**Pull request closed without merging**